dygraph/ppdet/modeling/proposal_generator/rpn_head.py

```diff
@@ -0,0 +1,221 @@
+        self.test_proposal = test_proposal
+        if isinstance(anchor_generator, dict):
+            self.anchor_generator = AnchorGenerator(**anchor_generator)
+        if isinstance(rpn_target_assign, dict):
+            self.rpn_target_assign = RPNTargetAssign(**rpn_target_assign)
+        if isinstance(train_proposal, dict):
+            self.train_proposal = ProposalGenerator(**train_proposal)
+        if isinstance(test_proposal, dict):
+            self.test_proposal = ProposalGenerator(**test_proposal)
+
+        num_anchors = self.anchor_generator.num_anchors
+        self.rpn_feat = RPNFeat(in_channel, in_channel)
+        # rpn head is shared with each level
+        # rpn roi classification scores
+        self.rpn_rois_score = nn.Conv2D(
+            in_channels=in_channel,
+            out_channels=num_anchors,
+            kernel_size=1,
+            padding=0,
+            weight_attr=paddle.ParamAttr(initializer=Normal(
+                mean=0., std=0.01)))
+
+        # rpn roi bbox regression deltas
+        self.rpn_rois_delta = nn.Conv2D(
+            in_channels=in_channel,
+            out_channels=4 * num_anchors,
+            kernel_size=1,
+            padding=0,
+            weight_attr=paddle.ParamAttr(initializer=Normal(
+                mean=0., std=0.01)))
+
+    @classmethod
+    def from_config(cls, cfg, input_shape):
+        # FPN share same rpn head
+        if isinstance(input_shape, (list, tuple)):
+            input_shape = input_shape[0]
+        return {'in_channel': input_shape.channels}
+
+    def forward(self, feats, inputs):
+        rpn_feats = self.rpn_feat(feats)
+        scores = []
+        deltas = []
+
+        for rpn_feat in rpn_feats:
+            rrs = self.rpn_rois_score(rpn_feat)
+            rrd = self.rpn_rois_delta(rpn_feat)
+            scores.append(rrs)
+            deltas.append(rrd)
+
+        anchors = self.anchor_generator(rpn_feats)
+
+        rois, rois_num = self._gen_proposal(scores, deltas, anchors, inputs)
+
+        if self.training:
+            loss = self.get_loss(scores, deltas, anchors, inputs)
+            return rois, rois_num, loss
+        else:
+            return rois, rois_num, None
+
+    def _gen_proposal(self, scores, bbox_deltas, anchors, inputs):
+        """
+        scores (list[Tensor]): Multi-level scores prediction 
+        bbox_deltas (list[Tensor]): Multi-level deltas prediction
+        anchors (list[Tensor]): Multi-level anchors 
+        inputs (dict): ground truth info
+        """
+        prop_gen = self.train_proposal if self.training else self.test_proposal
+        im_shape = inputs['im_shape']
+        batch_size = im_shape.shape[0]
+        rpn_rois_list = [[] for i in range(batch_size)]
+        rpn_prob_list = [[] for i in range(batch_size)]
+        rpn_rois_num_list = [[] for i in range(batch_size)]
+        for rpn_score, rpn_delta, anchor in zip(scores, bbox_deltas, anchors):
+            for i in range(batch_size):
```

Comment:
add comments for why not  do batch-computing for prop_gen?

Reply:
Done